lib/ast/expression.ts

```diff
@@ -1191,3 +1191,88 @@ export class ChainExpression extends Expression {
         return [null, newScope];
     }
 }
+
+
+export class JoinExpression extends Expression {
+    lhs : Expression;
+    rhs : Expression;
+    condition : BooleanExpression|null;
+
+    constructor(location : SourceRange|null,
+                left : Expression,
+                right : Expression,
+                condition : BooleanExpression|null, 
+                schema : FunctionDef|null) {
+        super(location, schema);
+
+        this.lhs = left;
+        this.rhs = right;
+        this.condition = condition;
+    }
+
+    get priority() : SyntaxPriority {
+        return SyntaxPriority.Join;
+    }
+
+    toSource() : TokenStream {
+        if (this.condition) {
+            return List.concat(
+                addParenthesis(this.priority, this.lhs.priority, this.lhs.toSource()),
+                'join',
+                addParenthesis(this.priority, this.rhs.priority, this.rhs.toSource()),
+                'on',
+                addParenthesis(this.priority, this.condition.priority, this.condition.toSource())
+            );
+        } else {
+            return List.concat(
+                addParenthesis(this.priority, this.lhs.priority, this.lhs.toSource()),
+                'join',
+                addParenthesis(this.priority, this.rhs.priority, this.rhs.toSource())
+            );
+        }
+    }
+
+    equals(other : Expression) : boolean {
+        return other instanceof JoinExpression &&
+            this.lhs === other.lhs && this.rhs === other.rhs && this.condition === other.condition;
+    }
+
+    toLegacy(into_params : InputParam[] = [], scope_params : string[] = []) : legacy.Table {
+        throw new Error('Not supported by legacy ThingTalk');
+    }
+
+    visit(visitor : NodeVisitor) : void {
+        visitor.enter(this);
+        this.lhs.visit(visitor);
+        this.rhs.visit(visitor);
+        if (this.condition)
+            this.condition.visit(visitor);
+        visitor.exit(this);
+    }
+
+    clone() : JoinExpression {
+        return new JoinExpression(
+            this.location,
+            this.lhs,
+            this.rhs,
+            this.condition,
```

Comment:
clone() must be a deep-clone (see also #364 )

Reply:
fixed in 61a6e46.